File: Group-Video/OpenVideoCall-iOS-Objective-C/OpenVideoCall/VideoView.m
//
//  VideoView.m
//  OpenVideoCall
//
//  Created by GongYuhua on 2016/11/17.
//  Copyright © 2016年 Agora. All rights reserved.
//

#import "VideoView.h"

@interface VideoView ()
@property (weak, nonatomic) UIView *infoView;
@property (weak, nonatomic) UILabel *infoLabel;
@end

@implementation VideoView
- (void)setIsVideoMuted:(BOOL)isVideoMuted {
    _isVideoMuted = isVideoMuted;
    self.videoView.hidden = isVideoMuted;
}

- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        self.translatesAutoresizingMaskIntoConstraints = NO;
        self.backgroundColor = [UIColor whiteColor];
        
        [self addVideoView];
        [self addInfoView];
    }
    
    return self;
}

- (void)updateInfo:(MediaInfo *)info {
    self.infoLabel.text = info.description;
}

- (void)addVideoView {
    UIView *videoView = [[UIView alloc] init];
    videoView.translatesAutoresizingMaskIntoConstraints = NO;
    videoView.backgroundColor = [UIColor clearColor];
    [self addSubview:videoView];
    
    NSArray *videoViewH = [NSLayoutConstraint constraintsWithVisualFormat:@"H:|[video]|" options:0 metrics:nil views:@{@"video": videoView}];
    NSArray *videoViewV = [NSLayoutConstraint constraintsWithVisualFormat:@"V:|[video]|" options:0 metrics:nil views:@{@"video": videoView}];
    [NSLayoutConstraint activateConstraints:videoViewH];
    [NSLayoutConstraint activateConstraints:videoViewV];
    
    self.videoView = videoView;
}

- (void)addInfoView {
    UIView *infoView = [[UIView alloc] init];
    infoView.translatesAutoresizingMaskIntoConstraints = NO;
    infoView.backgroundColor = UIColor.clearColor;
    
    [self addSubview:infoView];
    self.infoView = infoView;
    

    NSArray *infoViewH = [NSLayoutConstraint constraintsWithVisualFormat: @"H:|[info]|" options: 0 metrics: nil views: @{@"info": infoView}];
    NSArray *infoViewV = [NSLayoutConstraint constraintsWithVisualFormat: @"V:|[info(==135)]" options: 0 metrics: nil views: @{@"info": infoView}];
    
    [NSLayoutConstraint activateConstraints:infoViewH];
    [NSLayoutConstraint activateConstraints:infoViewV];
    

    UILabel *infoLabel = [self createInfoLabel];
    [infoView addSubview:infoLabel];
    self.infoLabel = infoLabel;
    
    NSArray *labelV = [NSLayoutConstraint constraintsWithVisualFormat: @"V:|-(20)-[info]" options: 0 metrics: nil views: @{@"info": infoLabel}];
    NSArray *labelH = [NSLayoutConstraint constraintsWithVisualFormat: @"H:|-(10)-[info]" options: 0 metrics: nil views: @{@"info": infoLabel}];
    [NSLayoutConstraint activateConstraints:labelV];
    [NSLayoutConstraint activateConstraints:labelH];
}

- (UILabel *)createInfoLabel {
    UILabel *label = [[UILabel alloc] init];
    label.translatesAutoresizingMaskIntoConstraints = NO;
    
    label.text = @" ";
    label.shadowOffset = CGSizeMake(0, 1);
    label.shadowColor = UIColor.blackColor;
    label.numberOfLines = 0;
    label.font = [UIFont systemFontOfSize:12];
    label.textColor = UIColor.whiteColor;
    
    return label;
}

@end

File: Group-Video/OpenVideoCall-iOS/OpenVideoCall/VideoView.swift
//
//  VideoView.swift
//  OpenVideoCall
//
//  Created by GongYuhua on 2/14/16.
//  Copyright © 2016 Agora. All rights reserved.
//

import UIKit

class VideoView: UIView {
    
    fileprivate(set) var videoView: UIView!
    
    fileprivate var infoView: UIView!
    fileprivate var infoLabel: UILabel!
    
    var isVideoMuted = false {
        didSet {
            videoView?.isHidden = isVideoMuted
        }
    }
    
    override init(frame frameRect: CGRect) {
        super.init(frame: frameRect)
        translatesAutoresizingMaskIntoConstraints = false
        backgroundColor = UIColor.white
        
        addVideoView()
        addInfoView()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

extension VideoView {
    func update(with info: StatisticsInfo) {
        infoLabel?.text = info.description()
    }
}

private extension VideoView {
    func addVideoView() {
        videoView = UIView()
        videoView.translatesAutoresizingMaskIntoConstraints = false
        videoView.backgroundColor = UIColor.clear
        addSubview(videoView)
        
        let videoViewH = NSLayoutConstraint.constraints(withVisualFormat: "H:|[video]|", options: [], metrics: nil, views: ["video": videoView])
        let videoViewV = NSLayoutConstraint.constraints(withVisualFormat: "V:|[video]|", options: [], metrics: nil, views: ["video": videoView])
        NSLayoutConstraint.activate(videoViewH + videoViewV)
    }
    
    func addInfoView() {
        infoView = UIView()
        infoView.translatesAutoresizingMaskIntoConstraints = false
        infoView.backgroundColor = UIColor.clear
        
        addSubview(infoView)
        let infoViewH = NSLayoutConstraint.constraints(withVisualFormat: "H:|[info]|", options: [], metrics: nil, views: ["info": infoView])
        let infoViewV = NSLayoutConstraint.constraints(withVisualFormat: "V:[info(==140)]|", options: [], metrics: nil, views: ["info": infoView])
        NSLayoutConstraint.activate(infoViewH + infoViewV)
        
        func createInfoLabel() -> UILabel {
            let label = UILabel()
            label.translatesAutoresizingMaskIntoConstraints = false
            
            label.text = " "
            label.shadowOffset = CGSize(width: 0, height: 1)
            label.shadowColor = UIColor.black
            label.numberOfLines = 0
            
            label.font = UIFont.systemFont(ofSize: 12)
            label.textColor = UIColor.white
            
            return label
        }
        
        infoLabel = createInfoLabel()
        infoView.addSubview(infoLabel)
        
        let top: CGFloat = 20
        let left: CGFloat = 10
        
        let labelV = NSLayoutConstraint.constraints(withVisualFormat: "V:|-(\(top))-[info]", options: [], metrics: nil, views: ["info": infoLabel])
        let labelH = NSLayoutConstraint.constraints(withVisualFormat: "H:|-(\(left))-[info]", options: [], metrics: nil, views: ["info": infoLabel])
        NSLayoutConstraint.activate(labelV)
        NSLayoutConstraint.activate(labelH)
    }
}
